### PR TITLE
[SIM_FLUTTER-60] [BE] Create Get Accounts API

### PR DESCRIPTION
**Pull request by IanUrriza10**
### Issue Links
- [backlog](https://framgiaph.backlog.com/view/SIM_FLUTTER-60)
- [slack](https://sun-philippine.slack.com/archives/C059L7XAXFB/p1687246980761349)
###  Notes
- to create accounts, do `php artisan db:seed --class=UserSeeder`
- then do `php artisan db:seed --class=AccountSeeder`
- visit URLs
    - http://127.0.0.1:8000/api/accounts
    - http://127.0.0.1:8000/api/accounts/{account_id}
    - http://127.0.0.1:8000/api/users/{user_id}/accounts

- For Testing
    - Use login first to get a `user_id` 
    - All calls are in Accounts Collection
    - `/accounts/{{account_id}}` needs an account_id first
        - either put there manually or call `/api/accounts` to set the first account id 
### Definition of done
- [x] create api/accounts route
- [x] display all users from the database
- [x] display all accounts of a specific user

### Screenshots
image
image
image

### CODE REVIEW

File: backend/app/Http/Controllers/AccountController.php [hunk 1]
```diff
@@ -2,7 +2,18 @@
...
-    //
+    public function index()
+    {
+        return AccountResource::collection(Account::all());
+    }
+
+    public function show(Account $account)
+    {
+        return AccountResource::make(($account));
+    }
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AccountController.php]: was this checked if this was working?

wrong parameters for the show function
should just be $id

File: backend/app/Http/Controllers/UserController.php [hunk 2]
```diff
@@ -36,4 +36,12 @@ public function login(Request $request)
 
         return LoginResource::make(array_merge($user->toArray(), ['token' => $token]));
     }
+    public function myAccounts(Request $request)
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/UserController.php]: space 
```suggestion
    }
    
    public function myAccounts(Request $request)
```

File: backend/routes/api.php [hunk 3]
```diff
@@ -23,4 +24,7 @@
         return UserResource::make($request->user());
     });
     Route::apiResource('users', UserController::class)->only(['index']);
+    Route::get('/users/{user_id}/accounts', [UserController::class, 'myAccounts']);
```
Comment (jstephend-sun) [inline backend/routes/api.php]: refer to https://laravel.com/docs/8.x/controllers#restful-nested-resources

for nested routing

this should be another apiResource, and function should be index 
ex Route::apiResource('users.accounts', UserAccountController::class)->only('index')

File: backend/app/Models/Transaction.php [hunk 4]
```diff
@@ -9,6 +9,10 @@ class Transaction extends Model
 {
     use HasFactory;
 
+    protected $casts = [
+        'id' => 'string',
+    ];
+
```
Comment (jstephend-sun) [inline backend/app/Models/Transaction.php]: as i found out later, this is not enough to handle the model's id being uuid

refer to #13

1. copy trait file to this PR as well  `backend/app/Traits/Uuid.php`
2. do the same as on pr13, `backend/app/Models/User.php:13` (right side of diff)
3. remove this `protected $casts`
4. and add same lines from pr13 as well `backend/app/Models/User.php:41-55` (right side of diff)

File: backend/app/Models/Account.php [hunk 5]
```diff
@@ -9,6 +9,14 @@ class Account extends Model
 {
     use HasFactory;
 
+    protected $appends = [ //For adding custom logic
+        'balance'
+    ];
+
+    protected $casts = [
```
Comment (jstephend-sun) [inline backend/app/Models/Account.php]: do the same here with pr 13 trait

File: backend/app/Models/Account.php [hunk 6]
```diff
@@ -9,6 +9,14 @@ class Account extends Model
 {
     use HasFactory;
 
+    protected $appends = [ //For adding custom logic
```
Comment (jstephend-sun) [inline backend/app/Models/Account.php]: add comments same with the documentation here, you can also reference that section

File: backend/app/Models/Account.php [hunk 7]
```diff
@@ -18,4 +26,17 @@ public function accountTransactions()
...
+
+    //Takes the latest transaction and outputs the sum of balance and transaction amount if exists
+    //otherwise returns null
+    public function getBalanceAttribute()
+    {
+        $latest_transaction = $this->accountTransactions()->latest()->first();
+
+        if (!$latest_transaction) {
+            return null;
+        }
```
Comment (jstephend-sun) [inline backend/app/Models/Account.php]: should this be null or `0`?

File: backend/app/Http/Controllers/UserController.php [hunk 8]
```diff
@@ -36,4 +36,12 @@ public function login(Request $request)
...
+    {
+        $user_id = $request->user_id;
+
+        $user = User::findOrFail($user_id);
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/UserController.php]: current logged in user can be easily retrieved by
Auth::user(),

then for id Auth::id()

File: backend/routes/api.php (hunk 3 above)
Comment (jstephend-sun) [reply backend/routes/api.php]: you need to create a pivot controller here
ex. pivot/UserAccountController

File: backend/app/Models/Account.php (hunk 6 above)
Comment (jstephend-sun) [inline backend/app/Models/Account.php]: property $fillable is missing, include it

File: backend/app/Http/Resources/AccountResource.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+     * @param  \Illuminate\Http\Request  $request
+     * @return array|\Illuminate\Contracts\Support\Arrayable|\JsonSerializable
+     */
+    public function toArray($request)
+    {
+
+        $data = parent::toArray($request);
+
+        $data['amount'] = $this->getBalanceAttribute();
+        $data['links'] = [
```
Comment (jstephend-sun) [inline backend/app/Http/Resources/AccountResource.php]: property links not needed, theres no mention of implementing HATEOAS pattern yet

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #18

review time: 20mins
comments: 10

File: backend/routes/api.php (hunk 3 above)
Comment (jstephend-sun) [reply backend/routes/api.php]: pivot may not be the right term haha 

can be `nested` instead

File: backend/app/Http/Controllers/AccountController.php (hunk 1 above)
Comment (IanUrriza10) [reply backend/app/Http/Controllers/AccountController.php]: image

Comment (IanUrriza10) [reply backend/app/Http/Controllers/AccountController.php]: I've moved it out of the auth middleware to test but i don't think it affects it

Comment (jstephend-sun) [reply backend/app/Http/Controllers/AccountController.php]: ok, just remove the double parenthesis

File: backend/app/Http/Controllers/UserController.php (hunk 2 above)
Comment (jstephend-sun) [reply backend/app/Http/Controllers/UserController.php]: @IanUrriza10 pls also change the function name, `myAccounts` doesnt seem to fit with other relationship function names

File: backend/routes/api.php [hunk 10]
```diff
@@ -24,7 +25,7 @@
         return UserResource::make($request->user());
     });
     Route::apiResource('users', UserController::class)->only(['index']);
-    Route::get('/users/{user_id}/accounts', [UserController::class, 'myAccounts']);
+    Route::resource('users.accounts',UserController::class)->shallow();
```
Comment (jstephend-sun) [inline backend/routes/api.php]: use apiResource, add only() chain

this is exposing unhandled routes
image

Comment (jstephend-sun) [inline backend/routes/api.php]: also, its using UserController? why?

the nested resource is Accounts, i expect it to be Account

image

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #18

review time: 5mins
comments: 2

**Review by Dbulawan13 — DISMISSED** [review]
Approve PR #18 

Working:
- Able to get accounts via
   - all accounts
   - user_id
   - accounts data

https://github.com/framgia/sph-flutter/assets/86758474/dc9304ba-0beb-4615-a7e2-a70a85775b67

Testing Hours: 1
Comment: 0
Iteration: 0

**Review by jstephend-sun — APPROVED** [review]
Approved #18

review time: 5mins
comments: 2

**Pull request merged**